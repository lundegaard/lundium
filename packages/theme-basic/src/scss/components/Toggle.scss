@import '../core';

/*
 -  Toggle
 -  (<input type="checkbox" class="form-control" />)
 */


.form-group--toggle {
	.control-label {
		height: $toggle-height;
		width: $toggle-width;
		margin-block-end: 0;
		background-color: $theme-green;
		border-radius: ( $toggle-height * .5 );
		color: $toggle-color;
		font-weight: 700;
		font-size: $toggle-font-size;
		text-transform: uppercase;
		display: inline-flex;
		align-items: center;
		position: relative;
	}

	.form-control {
		position: absolute;
		height: auto;
		z-index: -1;
		opacity: 0;

		& ~ .control-label {
			background-color: $theme-red;
			justify-content: flex-end;
			padding-inline-end: ( $toggle-icon-size + 10px );
			padding-inline-start: 0;
			transition:
				all $transition-speed $transition-easing;
			cursor: pointer;
			user-select: none;
		}

		&:focus ~ .control-label {
			box-shadow: $input-focus-box-shadow;
		}

		& ~ .control-label::before {
			@include icon(status-rejected);
			position: absolute;
			inset-block-start: $toggle-icon-offset;
			inset-inline: auto $toggle-icon-offset;
			display: block;
			font-size: $toggle-icon-size;
			color: $toggle-icon-color;
			transition: all $transition-speed $transition-easing;
			transform: translateX(0);
		}

		&:checked ~ .control-label {
			padding-inline-start: ( $toggle-icon-size + 10px );
			padding-inline-end: 0;
			background-color: $toggle-true-bg;
			justify-content: flex-start;
		}

		&:checked ~ .control-label::before {
			@include icon-content(status-success);
			inset-block-start: $toggle-icon-offset;
			transform: translateX( -($toggle-width - $toggle-icon-size - $toggle-icon-offset * 3) );

			&:dir(rtl) {
				transform: translateX( $toggle-width - $toggle-icon-size - $toggle-icon-offset * 3 );
			}
		}

		&:disabled,
		&[disabled] {
			& ~ .control-label {
				background-color: $theme-gray-400;
				color: $white;
			}
		}
	}
}