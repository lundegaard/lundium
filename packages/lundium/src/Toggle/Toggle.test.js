import React from 'react';
import { mount, shallow } from 'enzyme';

import Toggle from './Toggle';

describe('<Toggle />', () => {
	const checkedLabel = 'Open';
	const uncheckedLabel = 'Closed';
	const name = 'generic-toggle';

	const requiredProps = {
		checkedLabel,
		uncheckedLabel,
		name,
	};

	const checkboxSelector = 'input[type="checkbox"]';
	const labelSelector = `label[htmlFor^="${name}"]`;

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Toggle />);
		expect(wrapper).toMatchSnapshot();
	});

	it('is checked when receives prop checked={true} and has correct label', () => {
		const wrapper = mount(<Toggle {...requiredProps} checked readOnly />);
		const checkbox = wrapper.find(checkboxSelector);
		const label = wrapper.find(labelSelector);
		expect(label.text()).toBe(checkedLabel);
		expect(checkbox.props().checked).toBe(true);
	});

	it('is not checked when does not receive prop checked and has correct label', () => {
		const wrapper = mount(<Toggle {...requiredProps} />);
		const checkbox = wrapper.find(checkboxSelector);
		const label = wrapper.find(labelSelector);
		expect(checkbox.props().checked).toBeFalsy();
		expect(label.text()).toBe(uncheckedLabel);
	});

	it('is not checked when receives prop checked={false} and has correct label', () => {
		const wrapper = mount(
			<Toggle {...requiredProps} checked={false} readOnly />,
		);
		const checkbox = wrapper.find(checkboxSelector);
		const label = wrapper.find(labelSelector);
		expect(checkbox.props().checked).toBeFalsy();
		expect(label.text()).toBe(uncheckedLabel);
	});
});
