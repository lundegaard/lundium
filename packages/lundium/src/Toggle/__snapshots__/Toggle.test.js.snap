// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> matches snapshot when shallowly rendered 1`] = `
<forwardRef(FormField)
  control={[Function]}
  kind="toggle"
/>
`;
